src/main/java/org/opentripplanner/updater/spi/TimetableSnapshotFlushUpdater.java

```diff
@@ -0,0 +1,66 @@
+package org.opentripplanner.updater.spi;
+
+import java.time.Duration;
+import org.opentripplanner.ext.siri.SiriTimetableSnapshotSource;
+import org.opentripplanner.updater.TimetableSnapshotSourceParameters;
+import org.opentripplanner.updater.trip.TimetableSnapshotSource;
+
+/**
+ * Updater that flushes the snapshot buffer by committing pending changes.
+ * The updater runs at the same frequency as the timetable snapshot throttle frequency. See
+ * {@link org.opentripplanner.updater.TimetableSnapshotSourceParameters#withMaxSnapshotFrequency(Duration)}
+ * It means that pending changes are committed at least once per throttling period.
+ * This is useful for a system where real-time updates are sent in batches followed by silent periods.
+ * In this case flushing the buffer periodically guarantees that a throttled update will be applied timely.
+ * In a system that receives a steady flow of updates, this updater is not required but does not
+ * hurt performance either.
+ */
+public class TimetableSnapshotFlushUpdater extends PollingGraphUpdater {
```

Comment:
This approach with a PollingGraphUpdater seems like it would work in most cases, but it's slightly different than what I had in mind. Today we talked about the possibility of tasks piling up in the queue due to slow indexing steps, and I mentioned a different way of scheduling the snapshot flush task. What I had in mind was scheduling it directly on `GraphUpdaterManager.scheduler` which is a `ScheduledExecutorService` so has the `scheduleWithFixedDelay` method [(javadoc link)](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay-java.lang.Runnable-long-long-java.util.concurrent.TimeUnit-). This should ensure a snapshot task runs at least N seconds after the previous one completes, with plenty of intervening time for realtime update tasks to run.

Reply:
And so far it does look to me like things would work as desired if this single repeating task were the only thing that causes `purgeAndCommit()` to be called.

Reply:
> This approach with a PollingGraphUpdater seems like it would work in most cases, but it's slightly different than what I had in mind. Today we talked about the possibility of tasks piling up in the queue due to slow indexing steps, and I mentioned a different way of scheduling the snapshot flush task. What I had in mind was scheduling it directly on `GraphUpdaterManager.scheduler` which is a `ScheduledExecutorService` so has the `scheduleWithFixedDelay` method [(javadoc link)](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay-java.lang.Runnable-long-long-java.util.concurrent.TimeUnit-). This should ensure a snapshot task runs at least N seconds after the previous one completes, with plenty of intervening time for realtime update tasks to run.

This is indeed much simpler and much clearer  to schedule the task directly in the executor service.
I have updated the code accordingly.

Reply:
> And so far it does look to me like things would work as desired if this single repeating task were the only thing that causes `purgeAndCommit()` to be called.

I agree, I will try to adapt the code accordingly. It requires some changes in the integration testing framework since the new snapshot becomes available asynchronously.

Reply:
I think those changes are small since you can call `flushBuffer` right after applying the updates.

Reply:
You're right, there are few places in the test code where the snapshot buffer needs to be flushed.
Some tests verify the throttling mechanism and become obsolete.
I have updated the code accordingly.